src/gui/aboutdialog.h

```diff
@@ -46,4 +48,5 @@ class AboutDialog : public QDialog
 
 private:
     Ui::AboutDialog *m_ui;
+    CachedSettingValue<QSize> m_storeDialogSize;
```

Comment:
IMO, storing it in class field is redundant. All you need is to use it once when initializing the dialog, and then save the current value when closing.

Reply:
> IMO, storing it in class field is redundant.

I can move it to anonymous namespace, but then the variable will be in memory all the time which is not we want.

>All you need is to use it once when initializing the dialog, and then save the current value when closing.

Yes, in the past I proposed a very thin wrapper for this purpose, however that class was deemed superseded by `CachedSettingValue` class. Now we only have `CachedSettingValue` available...
Using `SettingsStorage` for this purpose is possible but it would be not as clean as using `CachedSettingValue`, e.g. the type `QSize` and `key` would be repeated in 2 instances separately.

Reply:
>Using `SettingsStorage` for this purpose is possible

That's exactly what I meant.
>the type `QSize` and `key` would be repeated in 2 instances (load & store) separately.

Well, if that's the problem...
>Yes, in the past I proposed a very thin wrapper for this purpose, however that class was deemed superseded by `CachedSettingValue` class.

I can't remember what it was... please remind me of its concept.

Reply:
>I can't remember what it was... please remind me of its concept.

I faintly recall that it is similar to `CachedSettingValue` but without the `T m_value` member variable. Ideally I would also try to avoid having `QString m_keyName` variable although I haven't got an idea yet.
IMO `CachedSettingValue` without `T m_value` would be the optimal wrapper class for this kind of usage.

For the usage here, IMO `CachedSettingValue` is usable albeit a little more heavier.

Reply:
>I faintly recall that it is similar to `CachedSettingValue` but without the `T m_value` member variable.

Well, it seems useful to encapsulate the setting name and its type without saving its value, it can save a couple of bytes in cases where intermediate access is not required (like in these dialogs).
But I don't think "sheepskin is worth the trouble" if its implementation will duplicate most of `CachedSettingValue` code. Ideally, `CachedSettingValue` should inherits non cached `SettingValue` class.

>Ideally I would also try to avoid having `QString m_keyName` variable although I haven't got an idea yet.

I can't even imagine what it could be... so that it brings some benefit.

>For the usage here, IMO CachedSettingValue is usable albeit a little more heavier.

Ok.

Reply:
>Ideally, CachedSettingValue should inherits non cached SettingValue class.

Luckily I was able to come up a workable implementation. IMO it is not that bad at all:
```c++
template <typename T>
class SettingValue
{
public:
    explicit SettingValue(const char *keyName)
        : m_keyName {QLatin1String(keyName)}
    {
    }

    T value() const
    {
        return loadValue();
    }

    operator T() const
    {
        return value();
    }

    SettingValue<T> &operator=(const T &value)
    {
        storeValue(value);
        return *this;
    }

private:
    // regular load/save pair
    template <typename U, typename std::enable_if_t<!std::is_enum<U>::value, int> = 0>
    U loadValue()
    {
        return SettingsStorage::instance()->loadValue(m_keyName).template value<U>();
    }

    template <typename U, typename std::enable_if_t<!std::is_enum<U>::value, int> = 0>
    void storeValue(const U &value)
    {
        SettingsStorage::instance()->storeValue(m_keyName, value);
    }

    // load/save pair for an enum
    // saves literal value of the enum constant
    template <typename U, typename std::enable_if_t<std::is_enum<U>::value, int> = 0>
    U loadValue()
    {
        return Utils::String::toEnum(loadValue<QString>());
    }

    template <typename U, typename std::enable_if_t<std::is_enum<U>::value, int> = 0>
    void storeValue(const U value)
    {
        storeValue(Utils::String::fromEnum(value));
    }

    const QString m_keyName;
};


template <typename T>
class CachedSettingValue
{
public:
    explicit CachedSettingValue(const char *keyName, const T &defaultValue = {})
        : m_setting {keyName}
        , m_cache {loadValue(keyName, defaultValue)}
    {
    }

    // The signature of the ProxyFunc should be equivalent to the following:
    // T proxyFunc(const T &a);
    template <typename ProxyFunc>
    explicit CachedSettingValue(const char *keyName, const T &defaultValue
                                , ProxyFunc &&proxyFunc)
        : m_setting {keyName}
        , m_cache {proxyFunc(loadValue(keyName, defaultValue))}
    {
    }

    T value() const
    {
        return m_cache;
    }

    operator T() const
    {
        return value();
    }

    CachedSettingValue<T> &operator=(const T &value)
    {
        if (m_cache == value)
            return *this;

        m_cache = value;
        m_setting = value;
        return *this;
    }

private:
    // regular load
    template <typename U, typename std::enable_if_t<!std::is_enum<U>::value, int> = 0>
    U loadValue(const char *keyName, const U &defaultValue)
    {
        return SettingsStorage::instance()->loadValue(QLatin1String(keyName), defaultValue).template value<U>();
    }

    // load function for an enum
    template <typename U, typename std::enable_if_t<std::is_enum<U>::value, int> = 0>
    U loadValue(const char *keyName, const U defaultValue)
    {
        return Utils::String::toEnum(loadValue<QString>(keyName, {}), defaultValue);
    }

    SettingValue<T> m_setting;
    T m_cache;
};

// Also provide default value for `defaultValue`
// template <typename T, typename std::enable_if_t<std::is_enum<T>::value, int> = 0>
// T toEnum(const QString &serializedValue, const T &defaultValue = {})
```

I ran out of time today, I'll continue it later...

Reply:
>Luckily I was able to come up a workable implementation. IMO it is not that bad at all

I was talking about inheritance, in case you haven't noticed...
In addition, it loses many advantages, such as the "custom" default value, the preprocessing of the loaded value... So its scope will be small.

Reply:
>I was talking about inheritance, in case you haven't noticed...

I don't think inheritance would fit here, we can try to fit it but it would look weird and feel unnatural.
I intend `SettingValue` to be a thin/handy layer over `SettingsStorage` and nothing else.

>In addition, it loses many advantages, such as the "custom" default value, the preprocessing of the loaded value... So its scope will be small.

Those aren't needed for GUI widgets. Look in torrent creator dialog, most (if not all) instances of `CachedSettingValue` can be replaced by `SettingValue`. It would be widely adopted in GUI for this kind of use case.

Reply:
Okay, go ahead with it.
There is really no need to try to make it universal if we end up with it being universally inconvenient.

Reply:
Let's consider PR #14085 first. I'll come back to this PR after that one is done.